### ISSUE TITLE
Webpack Issues with util.ts pushdata() function

### ISSUE THREAD

**Issue by christroutner** [simpleledger/slp-metadatamaker.js#1]
There seems to be a typing issue when slp-mdm is compiled by webpack into a front-end application. I came across this error while trying integrate it into [wallet.fullstack.cash](https://wallet.fullstack.cash):

```
TypeError: "list" argument must be an Array of Buffers
    at Function.concat (index.js:399)
    at Object../node_modules/slp-mdm/dist/util.js.exports.pushdata (util.js:8)
    at Object../node_modules/slp-mdm/dist/util.js.exports.createOpReturnSend (util.js:152)
    at Object.send (tokentype1.js:7)
    at TokenType1.generateSendOpReturn (tokentype1.js:99)
    at Tokens.createTransaction (tokens.js:179)
    at Tokens.sendTokens (tokens.js:109)
    at MinimalBCHWallet.sendTokens (index.js:163)
    at _callee$ (send-tokens.js:114)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:274)
    at Generator.prototype.<computed> [as next] (runtime.js:97)
    at asyncGeneratorStep (asyncToGenerator.js:3)
    at _next (asyncToGenerator.js:25)
```

The error appears to be generated from [this pushdata function in the Util library](https://github.com/simpleledger/slp-metadatamaker.js/blob/master/lib/util.ts#L3-L21)

I ran into a similar issue working with Buffers. It was caused by a type error between webpack and typeforce, once I removed the type checking it worked fine. The polyfill used by webpack seems to fail a lot of type checking.

**Comment by christroutner** [comment]
Continuing my analysis:

- Here the [specific line](https://github.com/simpleledger/slp-metadatamaker.js/blob/master/lib/util.ts#L7) causing the error.

The error is actually being thrown within the browser by the `Buffer.concat()` function. Here is the working code in the browser:
```javascript
Buffer.concat = function concat (list, length) {
  if (!isArray(list)) {
    throw new TypeError('"list" argument must be an Array of Buffers')
  }

  if (list.length === 0) {
    return Buffer.alloc(0)
  }

  var i
  if (length === undefined) {
    length = 0
    for (i = 0; i < list.length; ++i) {
      length += list[i].length
    }
  }

  var buffer = Buffer.allocUnsafe(length)
  var pos = 0
  for (i = 0; i < list.length; ++i) {
    var buf = list[i]
    if (!Buffer.isBuffer(buf)) {
      throw new TypeError('"list" argument must be an Array of Buffers')
    }
    buf.copy(buffer, pos)
    pos += buf.length
  }
  return buffer
}
```

- Here is [reference documentation on Buffer.conat()](https://www.w3schools.com/nodejs/met_buffer_concat.asp)

**Comment by christroutner** [comment]
This is [the specific line](https://github.com/simpleledger/slp-metadatamaker.js/blob/master/lib/util.ts#L188) in `util.ts/createOpReturnSend()` that is causing the error. I found that by stepping through the code in the browser. It's a hard-coded value, so this is definitely a systemic issue and not bad data from the user.

- When the buffer is passed to `Buffer.concat()` it comes in as `[Uint8Array(1), Uint8Array(4)]` with values of `[04, 83, 76, 80, 0]`

- It's these specific lines within the `Buffer.concat()` function that is throwing the error:

```javascript
    var buf = list[i]
    if (!Buffer.isBuffer(buf)) {
      throw new TypeError('"list" argument must be an Array of Buffers')
    }
```

It's the first element in the array, the Uint8Array Buffer with a single value of 04 that is failing the `!Buffer.isBuffer(buf)` check.

I created this test code and verified the issue:

```javascript
console.log('Entering test code...')
      const buf = Uint8Array.from([4])
      const isBuf = Buffer.isBuffer(buf)
      console.log('isBuf: ', isBuf) // false
      console.log('...Exiting test code')
```

`isBuf` has a value of `false`. This is what is failing. I don't know enough about webpack, Uint8Array, or Buffer data types to know what the solution is.

**Comment by christroutner** [comment]
The code appears to be being evaluated correctly:

```javascript
export const pushdata = (buf: Buffer|Uint8Array): Buffer => {
  if (buf.length === 0) {
    return Buffer.from(Uint8Array.from([0x4C, 0x00]));
  } else if (buf.length < 0x4E) {
    return Buffer.concat([Uint8Array.from([buf.length]), buf]);
...
```
since `buf` can be *either* a Buffer or an Uint8Array, and a Uin8Array evaluates to `false` with `Buffer.isBuffer()`, it seems to me the issue is with the existing code. **Any data that comes through as a Uint8Array need to first be converted to a Buffer.**

**Looking for feedback on this conclusion.**

**Comment by christroutner** [comment]
Further research on this issue brought me to [this Stack Overflow thread](https://stackoverflow.com/questions/37228285/uint8array-to-arraybuffer). It looks like the solution to converting a Uint8Array in browser-land looks like this:

```javascript
function typedArrayToBuffer(array: Uint8Array): ArrayBuffer {
    return array.buffer.slice(array.byteOffset, array.byteLength + array.byteOffset)
}
```

**Comment by christroutner** [comment]
I tried integrating the above into the slp-mdm library without success.

The source of the error now appears to me to be a bad assumption that Uint8Arrays would evaluate as a Buffer.

I think in node.js land, everything comes through as Buffers and it's fine. But in Browser land, they come through as Uint8Arrays and they need to be converted a ArrayBuffer

**Comment by christroutner** [comment]
Talked with @blockparty-sh about this and he said that this browser-based code works fine:

```html
<script src='https://unpkg.com/slp-mdm'></script>
<script>
const buf2 = slpMdm.NFT1.Child.genesis(
  'IVAN', // symbol
  'gorilla',                                                           // name
  'https://fountainhead.cash/document.pdf',                           // document_uri
  'f8bf41177a5f5e808a7ccb648b51080b031f15ca8018d91a576263d6cc626eb6'  // document_hash
)
</script>
```

To figure out the difference, I think the same transaction would need to be constructed using both methods, then step through the code to figure out what is different between the two.

**Comment by christroutner** [comment]
I'm updating and closing this task. The issue seems to be isolated to webpack. If you compile your libraries with Browserify, this problem goes away. Webpack does not handle Buffers very well.

The solution I've adopted is to compile features into separate libraries with Browserify, publish them to npm, then call them on the front end using unpkg. This architecture is explained at a high-level in this video:
https://youtu.be/G7ptg7VIRnk?t=810

And I have more specific details in this video:
https://youtu.be/NGnuwLJWqZ4

**Comment by p0o** [comment]
I have the same [problem](https://github.com/simpleledger/slpjs/issues/41). I think this issue shouldn't have been closed with a workaround, I believe slp-mdm has problem with working in browsers and that's why packages like [buffer](https://www.npmjs.com/package/buffer) exist to fix the incompatibility between node js Buffer and browser Buffer. I tried adding it simply to this repo but TS is failing for some reason on build.

**Comment by blockparty-sh** [comment]
Happy to keep it open. As far as I know this is solely an issue with webpack and their retarded implementation, not an issue with mdm itself. Would be great if a solution could be found to fix this issue.

**Comment by p0o** [comment]
I'm not sure but I doubt it's only webpack. What I don't understand is how Buffer which is only native to Nodejs is used without a module import throughout the code and the browser based version up there somehow works? Is typescript polyfilling the Buffer?

**Comment by blockparty-sh** [comment]
https://github.com/feross/buffer

**Comment by christroutner** [comment]
I'd also love to hear if anyone has this issue with the new webpack.

But I can confirm that this is a webpack-specific issue, and it applies to Buffers generally. This issue doesn't have anything specifically to do with crypto or the slp-mdm library. I've seen webpack suffer from this issue on other projects that used node.js Buffers.

That's why I adopted a generic workflow of building node.js libraries with Browserify, publishing them as npm packages, and then using them on the frontend as external unpkg libraries. This was a general solution to any code that makes use of node.js Buffers.

**Comment by blockparty-sh** [comment]
Just tested and merged fix from James and published 0.0.7. Thanks. Feel free to re-open if similar issue happens with webpack.
